tests/test_genetic_value.py

```diff
@@ -209,79 +264,178 @@ def test_diff_ind_tree(self):
+            {
+                "site_id": [0],
+                "effect_size": [freqdep(-1, 1 / 2)],
+                "trait_id": [0],
+                "causal_state": ["T"],
+            }
+        )
+        genetic_df = pd.DataFrame(
             {
                 "trait_id": [0, 0, 0],
                 "individual_id": [0, 1, 2],
-                "genetic_value": [0, 11, 12],
+                "genetic_value": [0, freqdep(-1, 1 / 2), 2 * freqdep(-1, 1 / 2)],
             }
         )
 
-        pd.testing.assert_frame_equal(genetic_df, result_df, check_dtype=False)
+        pd.testing.assert_frame_equal(
+            effect_size_df, genetic_result.effect_size, check_dtype=False
+        )
+        pd.testing.assert_frame_equal(
+            genetic_df, genetic_result.genetic, check_dtype=False
+        )
 
     def test_triploid(self, sample_df):
         trait_df = pd.DataFrame(
             {
-                "trait_id": [0, 0],
-                "site_id": [0, 1],
-                "causal_state": ["T", "C"],
-                "effect_size": [1, 10],
+                "site_id": [0],
+                "effect_size": [1],
+                "trait_id": [0],
             }
         )
 
         ts = triploid_tree()
         tree = ts.first()
-        genetic = tstrait.GeneticValue(ts, sample_df)
+        genetic = tstrait.GeneticValue(ts, sample_df, alpha=-1, random_seed=1)
         g0 = genetic._individual_genotype(tree, ts.site(0), "T", ts.num_nodes)
         g1 = genetic._individual_genotype(tree, ts.site(1), "C", ts.num_nodes)
 
         np.testing.assert_equal(g0, np.array([1, 2]))
         np.testing.assert_equal(g1, np.array([1, 1]))
 
-        genetic_df = tstrait.genetic_value(ts=ts, trait_df=trait_df)
+        c0 = genetic._obtain_allele_count(tree, ts.site(0))
+        c1 = genetic._obtain_allele_count(tree, ts.site(1))
+
+        assert c0 == {"T": 3}
+        assert c1 == {"C": 2, "T": 1}
 
-        result_df = pd.DataFrame(
+        genetic_result = tstrait.sim_genetic(
+            ts=ts, trait_df=trait_df, alpha=-1, random_seed=2
+        )
+        effect_size_df = pd.DataFrame(
+            {
+                "site_id": [0],
+                "effect_size": [freqdep(-1, 1 / 2)],
+                "trait_id": [0],
+                "causal_state": ["T"],
+            }
+        )
+        genetic_df = pd.DataFrame(
             {
                 "trait_id": [0, 0],
                 "individual_id": [0, 1],
-                "genetic_value": [11, 12],
+                "genetic_value": [freqdep(-1, 1 / 2), 2 * freqdep(-1, 1 / 2)],
             }
         )
 
-        pd.testing.assert_frame_equal(genetic_df, result_df, check_dtype=False)
+        pd.testing.assert_frame_equal(
+            effect_size_df, genetic_result.effect_size, check_dtype=False
+        )
+        pd.testing.assert_frame_equal(
+            genetic_df, genetic_result.genetic, check_dtype=False
+        )
+
+    def test_allele_freq_one(self, sample_trait_model):
```

Comment:
With https://github.com/tskit-dev/tstrait/pull/65#discussion_r1297128889, I examined the case in the site where there is only 1 state (reverse mutation completely cancelled out the mutation), and I got the desired output.